File: analyze_sm2025.mjs
import { readFileSync } from 'fs';

const data = readFileSync('SM_2025.json', 'utf8');
const raw = JSON.parse(data);
// Structure: [ section1, section2, ... ] where each section is [block, block, ...], each block has .projects
function* allProjects(data) {
  const sections = Array.isArray(data) ? data : [data];
  for (const section of sections) {
    if (!Array.isArray(section)) continue;
    for (const block of section) {
      if (typeof block !== 'object') continue;
      const projects = block.projects;
      if (Array.isArray(projects)) for (const p of projects) yield p;
      else if (block.project_id) yield block;
    }
  }
}

let totalCases = 0;
const scoreCounts = {};
let shortlistedCount = 0;
const shortlistByType = {};

for (const p of allProjects(raw)) {
  if (typeof p !== 'object') continue;
  totalCases++;

  const dt = p.diplom_text;
  if (dt) {
    shortlistedCount++;
    shortlistByType[dt] = (shortlistByType[dt] || 0) + 1;
  }

  const level2 = p.level2 || {};
  const marks = level2.marks_and_comments || level2.marks_comments || level2.marks_scores || {};
  const scores = [];
  for (const judgeId of Object.keys(marks)) {
    const j = marks[judgeId];
    if (typeof j !== 'object') continue;
    let t = j.total;
    if (t == null) continue;
    if (typeof t === 'string' && t.trim().toLowerCase() === 'is_my') continue;
    const v = parseFloat(t);
    if (!Number.isNaN(v)) scores.push(v);
  }
  if (scores.length > 0) {
    const avg = scores.reduce((a, b) => a + b, 0) / scores.length;
    const rounded = Math.round(avg * 10) / 10;
    scoreCounts[rounded] = (scoreCounts[rounded] || 0) + 1;
  }
}

const totalWithScore = Object.values(scoreCounts).reduce((a, b) => a + b, 0);
console.log('='.repeat(60));
console.log('SILVER MERCURY 2025 — Competition results summary');
console.log('='.repeat(60));
console.log('');
console.log('Total cases considered:     ', totalCases);
console.log('Cases with score (level 2):', totalWithScore);
console.log('Shortlisted (any diploma): ', shortlistedCount);
console.log('');
console.log('Shortlist by diploma type:');
for (const k of ['SHORTLIST', 'SILVER', 'GOLD']) {
  if (shortlistByType[k]) console.log('  ', k + ':', shortlistByType[k]);
}
console.log('');
console.log('Score distribution (average score per case, 1 decimal):');
console.log('-'.repeat(40));
for (const score of Object.keys(scoreCounts).sort((a, b) => parseFloat(a) - parseFloat(b))) {
  console.log('  Score', String(parseFloat(score)).padStart(4) + ':', String(scoreCounts[score]).padStart(4), 'cases');
}
console.log('-'.repeat(40));
console.log('  Total with score:', totalWithScore);

File: analyze_sm2025.py
#!/usr/bin/env python3
"""Analyze SM_2025.json - Silver Mercury 2025 competition results."""

import json
from collections import defaultdict

def main():
    with open("SM_2025.json", "r", encoding="utf-8") as f:
        data = f.read()

    # File might be wrapped in double array
    raw = json.loads(data)
    if isinstance(raw, list) and len(raw) > 0 and isinstance(raw[0], list):
        blocks = raw[0]  # list of nomination blocks
    else:
        blocks = raw if isinstance(raw, list) else [raw]

    total_cases = 0
    score_counts = defaultdict(int)  # rounded score -> count
    shortlisted_count = 0
    shortlist_by_type = defaultdict(int)  # SHORTLIST, SILVER, GOLD

    for block in blocks:
        if not isinstance(block, dict):
            continue
        projects = block.get("projects", [])
        for p in projects:
            if not isinstance(p, dict):
                continue
            total_cases += 1

            # Shortlist: diplom_text present
            dt = p.get("diplom_text")
            if dt:
                shortlisted_count += 1
                shortlist_by_type[dt] = shortlist_by_type.get(dt, 0) + 1

            # Score: from level2 judge totals (average per project)
            level2 = p.get("level2") or {}
            marks = level2.get("marks_and_comments") or level2.get("marks_comments") or level2.get("marks_scores") or {}
            scores = []
            for judge_id, judge_data in marks.items():
                if not isinstance(judge_data, dict):
                    continue
                t = judge_data.get("total")
                if t is None:
                    continue
                if isinstance(t, str) and t.strip().lower() == "is_my":
                    continue
                try:
                    scores.append(float(t))
                except (TypeError, ValueError):
                    pass
            if scores:
                avg = sum(scores) / len(scores)
                # Round to 1 decimal for distribution
                rounded = round(avg, 1)
                score_counts[rounded] += 1

    # Build sorted score distribution
    if not score_counts:
        # Fallback: maybe scores are elsewhere
        print("No level2 marks_comments/marks_scores found. Checking structure...")
    else:
        print("=" * 60)
        print("SILVER MERCURY 2025 — Competition results summary")
        print("=" * 60)
        print()
        print(f"Total cases considered:     {total_cases}")
        print(f"Cases with score (level 2): {sum(score_counts.values())}")
        print(f"Shortlisted (any diploma):  {shortlisted_count}")
        print()
        print("Shortlist by diploma type:")
        for k in ["SHORTLIST", "SILVER", "GOLD"]:
            if k in shortlist_by_type:
                print(f"  {k}: {shortlist_by_type[k]}")
        print()
        print("Score distribution (average score per case, 1 decimal):")
        print("-" * 40)
        for score in sorted(score_counts.keys()):
            print(f"  Score {score:4.1f}: {score_counts[score]:4d} cases")
        print("-" * 40)
        print(f"  Total with score: {sum(score_counts.values())}")

if __name__ == "__main__":
    import sys
    try:
        main()
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        import traceback
        traceback.print_exc()
        sys.exit(1)
